.github/workflows/wheels.yml

```
@@ -190,17 +197,35 @@ jobs:
 
     - run: make sdist
 
-    - uses: actions/upload-artifact@v3
+    - uses: actions/upload-artifact@v4
       with:
-        name: dist
+        name: sdist
         path: dist/*.tar.gz
 
-  success:
-    permissions:
-      contents: none
+  pypi-publish:
+    if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
     needs: [build, windows, sdist]
     runs-on: ubuntu-latest
-    name: Wheels Successful
+    name: Upload release to PyPI
+    environment:
+      name: release-pypi
+      url: https://pypi.org/p/Pillow
+    permissions:
+      id-token: write
     steps:
-      - name: Success
-        run: echo Wheels Successful
+      - uses: actions/download-artifact@v4
+      - name: Merge artifacts
+        run: |
+          for artifact in macOS_x86_64 \
+                          macOS_arm64 \
+                          manylinux2014_musllinux \
+                          manylinux_2_28 \
+                          windows_x86 \
+                          windows_x64 \
+                          windows_arm64 \
+                          sdist; do
+            mv $artifact/* .
+            rmdir $artifact
+          done
```

Let's try something like https://github.com/scikit-build/cmake-python-distributions/pull/438 / https://github.com/scientific-python/cookie/pull/335 to avoid repeating the names:

```suggestion
      - uses: actions/download-artifact@v4
        with:
          path: all

      - name: Merge files
        run: |
          mkdir dist
          mv all/*/* dist

```

---

Ok, I've pushed something similar.